docs: dynamic host volume specification (#24810)
hashicorp/nomad: website/content/docs/other-specifications/volume/host.mdx

description: Learn about the Volume specification used to create and register dynamic host volumes to Nomad.
---

# Dynamic Host Volume Specification

The Nomad volume specification defines the schema for creating and registering
dynamic host volumes using the [`volume create`][] and [`volume register`][]
commands and the [`PUT /v1/volume/host/create`][api_volume_create] and [`PUT
/v1/volume/host/register`][api_volume_register] API endpoints.

Some attributes are only be supported by specific operation, while others may
have a different meaning for each action, so read the documentation for each
attribute carefully. The section [Differences Between Create and
Register](#differences-between-create-and-register) provides a summary of the
differences.

The file may be provided as either HCL or JSON to the commands and as JSON to
the API.

## Volume Specification Parameters

- `capacity` `(string: <optional>)` - The size of a volume in bytes. either the
  physical size of a disk or a quota, depending on the plugin. This field must
  be between the `capacity_min` and `capacity_max` values unless they are
  omitted. Accepts human-friendly suffixes such as `"100GiB"`. Only supported

>   for volume registration.

Author: The parser for the volume spec ignores the field entirely for CSI.

  capacity, in bytes. The capacity of a volume may be the physical size of a
  disk, or a quota, depending on the plugin. The specific size of the resulting
  volume will be somewhere between `capacity_min` and `capacity_max`; the exact
  behavior is up to the plugin. If you want to specify an exact size, set
  `capacity_min` and `capacity_max` to the same value. Accepts human-friendly
  suffixes such as `"100GiB"`. Plugins that cannot restrict the size of volumes
  may ignore this field.

- `capacity_max` `(string: <optional>)` - Option for requesting a maximum
  capacity, in bytes. The capacity of a volume may be the physical size of a
  disk, or a quota, depending on the plugin. The specific size of the resulting
  volume will be somewhere between `capacity_min` and `capacity_max`; the exact
  behavior is up to the plugin. If you want to specify an exact size, set
  `capacity_min` and `capacity_max` to the same value. Accepts human-friendly
  suffixes such as `"100GiB"`. Plugins that cannot restrict the size of volumes
  may ignore this field.

- `capability` <code>([Capability][capability]: &lt;required&gt;)</code> -

>   Option for validating the capability of a volume.

Reviewer: All DHV capabilities are `single-node-*`, right? I don't see that mentioned here in the DHV spec page, nor the `capability` page.

Author: I missed that, yeah. I'll copy over what's done in https://github.com/hashicorp/nomad/pull/24828

  any user with `node:read` ACL, even across namespaces, so they should not be
  treated as sensitive values.

- `namespace` `(string: <optional>)` - The namespace of the volume. This field
  overrides the namespace provided by the `-namespace` flag or `NOMAD_NAMESPACE`
  environment variable. Defaults to `"default"` if unset.

- `node_id` `(string)` - A specific node where you would like the volume to be
  created (refer to the [volume placement](#volume-placement) section. Optional
  for volume creation but required for volume registration.

- `node_pool` `(string: <optional>)` - A specific node pool where you would like
  the volume to be created (refer to the [volume placement](#volume-placement)
  section. Optional for volume creation or volume registration. If you provide
  also `node_id`, the node must be in the provided `node_pool`.

- `parameters` <code>(map<string|string>:nil)</code> - An optional key-value map
  of strings passed directly to the plugin to configure the volume. The details
  of these parameters are specific to the plugin.

- `plugin_id` `(string)` - The ID of the [dynamic host volume

>   plugin][dhv_plugin] that manages this volume. Required for volume

Reviewer 2: FYI I'm not checking DHV links until I review the entire DHV docs set in the uber branch. :)

Several fields are set automatically by Nomad or the plugin when `volume create`
or `volume register` commands (or the equivalent APIs) are successful and you
should not set their values if they are not supported by the operation.

In **volume creation** you must set the [`plugin_id`](#plugin_id) field. The
[`capacity`](#capacity) and [`host_path`](#host_path) fields are ignored.

In **volume registration** you must set the [`node_id`](#node_id) and
[`host_path`](#host_path) fields. The [`constraint`](#constraint) field is
ignored. The [`node_pool`](#node_pool), [`capacity_max`](#capacity_max), and
[`capacity_min`](#capacity_min) fields are ignored but must be consistent if
set; the `node_pool` must match the node set by `node_id`, and the minimum
capacity must be less than the maximum capacity and actual capacity.

## Volume Placement

The `volume create` command will create the volume on a specific node. If
`node_id` is set, the volume will be created on that node. The `node_pool` must
be unset or match that node. Otherwise, the Nomad server will iterate over the
available nodes and place the volume on the first available node based on the

> following factors:

Author: Deliberately undefined but in practice ordered by ID.

* The node cannot already have a host volume with the same name.
* If `node_pool` is set, the selected node must be in that node pool.
* The node must meet the constraints defined by the `constraint` fields.

## Updating a Volume Definition

The `volume create` and `volume register` commands allow updating a volume
definition, if the [`id`](#id) field if set. But not all fields can be updated
after the volume is created:

> * The `name`, `type`, and `plugin_id` fields cannot be updated.

Author: That's a good point about the longer-term operations story. This text describes the current state of things... so if it's ok with you I'm going to add an item to the API Jira's checklist to address this and follow-up with a change in a separate PR.

> * The `name`, `type`, and `plugin_id` fields cannot be updated.

Author: https://github.com/hashicorp/nomad/pull/24851

* The `node_id` field cannot be updated. The `node_pool` field can only be

>   changed from empty to the node pool that matches the `node_id` field.

Author: I didn't think you _could_ change the pool a node is in, but from a quick look at the state store you totally can. We only ever "place" a volume once though, so if the node's pool changes the volume's doesn't, and we don't check it again unless you _also_ change the node pool in the volume spec, at which point it would return an error because we don't think that should be allowed.

We should probably do one of two things:
1. Update the node pool of all the volumes on a given node if the node pool of the node is updated.
2. Simply ignore the node pool of a volume after it's initially written

If we do (1) we'd get consistent state for queries, so it feels like we should probably do that. But that gets a little weird because we don't, for example, reschedule jobs when the node pool changes. If we do (2) the volume author doesn't have to worry about updating the volume spec. Maybe we should do both? Would that inconsistency be worse?

>   changed from empty to the node pool that matches the `node_id` field.

Author: https://github.com/hashicorp/nomad/pull/24851

* The `capacity_min` and `capacity_max` fields can be updated, and may increase
  the volume size if the plugin supports it. Expansion may or may not be
  possible while the volume is in use, again depending on the plugin.
  Reducing volume capacity is not allowed.
* The `capacity` field can be updated only for volume registration.
* The `constraint` fields can be updated but are ignored after the volume is created.
* The `capability` blocks can be added or removed, but only if the capability is

>   not currently in use by a mounted volume.

Author: Yeah we call the plugin RPC on any change even if it should be a no-op. We _could_ optimize that away but it might be nice for plugin authors if during development they still get the RPC.

Plugins don't really know anything about capabilities because that's a "claim time" attribute -- it's a function of how schedule allocations and mount to tasks. I would prefer we keep those separate unless or until we get a feature request otherwise. The client can already absolutely enforce capabilities without the help of the plugins, so that just adds an extra source of potential conflicting behavior. (Or by way of analogy with CSI: the plugin is responsible for `NodeStageVolume` and the client is responsible for `NodePublishVolume`.)

---
layout: docs

> page_title: Host Volume Specification

Reviewer 2: ```suggestion
page_title: Dynamic Host Volume Specification
```
change to match `#` page title

> description: Learn about the Volume specification used to create and register dynamic host volumes to Nomad.

Reviewer 2: ```suggestion
description: |-
  Learn about the Nomad dynamic host volume specification. Create and 
  register dynamic host volumes using the `volume create` and `volume 
  register` commands and the `PUT /v1/volume/host/create` and
  `PUT /v1/volume/host/register` API endpoints.  Define capacity, 
  capability, constraint, node, node pool, and parameters passed directly 
  to the plugin to configure the volume. Learn how volume creation and 
  registration are different. Additionally, learn how to place a volume on 
  specific nodes, update a volume, and expand a volume's capacity.
```

---

# Dynamic Host Volume Specification

> The Nomad volume specification defines the schema for creating and registering
> dynamic host volumes using the [`volume create`][] and [`volume register`][]
> commands and the [`PUT /v1/volume/host/create`][api_volume_create] and [`PUT
> /v1/volume/host/register`][api_volume_register] API endpoints.

Reviewer 2: ```suggestion
This page provides reference information for the Nomad dynamic host volume specification. Create and register
dynamic host volumes using the [`volume create`][] and [`volume register`][]
commands and the [`PUT /v1/volume/host/create`][api_volume_create] and [`PUT
/v1/volume/host/register`][api_volume_register] API endpoints.  Define capacity, capability, constraint, node, node pool, and parameters passed directly to the plugin to configure the volume.
Learn how volume creation and registration are different. Additionally, learn how to place a volume on specific nodes, update a volume, and expand a volume's capacity.
```

Some attributes are only be supported by specific operation, while others may
have a different meaning for each action, so read the documentation for each
attribute carefully. The section [Differences Between Create and
Register](#differences-between-create-and-register) provides a summary of the
differences.

The file may be provided as either HCL or JSON to the commands and as JSON to
the API.

## Volume Specification Parameters

> - `capacity` `(string: <optional>)` - The size of a volume in bytes. either the

Reviewer 2: ```suggestion
- `capacity` `(string: <optional>)` - The size of a volume in bytes. Either the
```

  physical size of a disk or a quota, depending on the plugin. This field must
  be between the `capacity_min` and `capacity_max` values unless they are
  omitted. Accepts human-friendly suffixes such as `"100GiB"`. Only supported
  for volume registration.

- `capacity_min` `(string: <optional>)` - Option for requesting a minimum
  capacity, in bytes. The capacity of a volume may be the physical size of a
  disk, or a quota, depending on the plugin. The specific size of the resulting

>   volume will be somewhere between `capacity_min` and `capacity_max`; the exact

Reviewer 2: ```suggestion
  volume is somewhere between `capacity_min` and `capacity_max`; the exact
```

  behavior is up to the plugin. If you want to specify an exact size, set
  `capacity_min` and `capacity_max` to the same value. Accepts human-friendly
  suffixes such as `"100GiB"`. Plugins that cannot restrict the size of volumes
  may ignore this field.

- `capacity_max` `(string: <optional>)` - Option for requesting a maximum
  capacity, in bytes. The capacity of a volume may be the physical size of a
  disk, or a quota, depending on the plugin. The specific size of the resulting

>   volume will be somewhere between `capacity_min` and `capacity_max`; the exact

Reviewer 2: ```suggestion
  volume is somewhere between `capacity_min` and `capacity_max`; the exact
```

  behavior is up to the plugin. If you want to specify an exact size, set
  `capacity_min` and `capacity_max` to the same value. Accepts human-friendly
  suffixes such as `"100GiB"`. Plugins that cannot restrict the size of volumes
  may ignore this field.

- `capability` <code>([Capability][capability]: &lt;required&gt;)</code> -
  Option for validating the capability of a volume.

- `constraint` <code>([Constraint][constraint]: &lt;optional&gt;)</code> - A

>   restriction on the eligible nodes where a volume can be created (refer to the
>   [volume placement](#volume-placement) section). You can provide multiple

Reviewer 2: ```suggestion
  restriction on the eligible nodes where a volume can be created. Refer to the 
  [volume placement](#volume-placement) section for details. You can provide multiple
```

  `constraint` blocks to add more constraints. Optional for volume creation and
  ignored for volume registration.

- `id` `(string: <optional>)` - The ID of a previously created volume to update
  via `volume create` or `volume register`. You should never set this field when
  initially creating or registering a volume, and you should only use the values
  returned from the Nomad API for the ID.

- `host_path` `(string)` - The path on disk where the volume exists. You should

>   set this only for volume registration. It will be ignored for volume creation.

Reviewer 2: ```suggestion
  set this only for volume registration. It is ignored for volume creation.
```

> - `name` `(string: <required>)` - The name of the volume, which will be used as

Reviewer 2: ```suggestion
- `name` `(string: <required>)` - The name of the volume, which is used as
```

  the [`volume.source`][volume_source] field in job specifications that claim
  this volume. Host volume names must be unique per node. Names are visible to
  any user with `node:read` ACL, even across namespaces, so they should not be
  treated as sensitive values.

- `namespace` `(string: <optional>)` - The namespace of the volume. This field
  overrides the namespace provided by the `-namespace` flag or `NOMAD_NAMESPACE`
  environment variable. Defaults to `"default"` if unset.

- `node_id` `(string)` - A specific node where you would like the volume to be

>   created (refer to the [volume placement](#volume-placement) section). Optional

Reviewer 2: ```suggestion
  created. Refer to the [volume placement](#volume-placement) section for details. Optional
```

  for volume creation but required for volume registration.

- `node_pool` `(string: <optional>)` - A specific node pool where you would like

>   the volume to be created (refer to the [volume placement](#volume-placement)
>   section). Optional for volume creation or volume registration. If you also

Reviewer 2: ```suggestion
  the volume to be created. Refer to the [volume placement](#volume-placement)
  section for details. Optional for volume creation or volume registration. If you also
```

  provide `node_id`, the node must be in the provided `node_pool`.

- `parameters` <code>(map<string|string>:nil)</code> - An optional key-value map
  of strings passed directly to the plugin to configure the volume. The details
  of these parameters are specific to the plugin.

- `plugin_id` `(string)` - The ID of the [dynamic host volume
  plugin][dhv_plugin] that manages this volume. Required for volume
  creation.

- `type` `(string: <required>)` - The type of volume. Must be `"host"` for
  dynamic host volumes.

## Differences Between Create and Register

Several fields are set automatically by Nomad or the plugin when `volume create`

> or `volume register` commands (or the equivalent APIs) are successful and you

Reviewer 2: ```suggestion
or `volume register` commands, or the equivalent APIs, are successful and you
```

should not set their values if they are not supported by the operation.

In **volume creation** you must set the [`plugin_id`](#plugin_id) field. The
[`capacity`](#capacity) and [`host_path`](#host_path) fields are ignored.

In **volume registration** you must set the [`node_id`](#node_id) and
[`host_path`](#host_path) fields. The [`plugin_id`](#plugin_id) and
[`constraint`](#constraint) fields are ignored. The [`node_pool`](#node_pool),
[`capacity_max`](#capacity_max), and [`capacity_min`](#capacity_min) fields are
ignored but must be consistent if set; the `node_pool` must match the node set
by `node_id`, and the minimum capacity must be less than the `capacity_max` and
`capacity`.

## Volume Placement

> The `volume create` command will create the volume on a single node. If
> `node_id` is set, the volume will be created on that node. The `node_pool` must
> be unset or match that node. Otherwise, the Nomad server will iterate over the
> available nodes and place the volume on the first available node based on the

Reviewer 2: ```suggestion
The `volume create` command creates the volume on a single node. If
`node_id` is set, the volume is created on that node. The `node_pool` must
be unset or match that node. Otherwise, the Nomad server iterates over the
available nodes and place the volume on the first available node based on the
```

[`host_path`](#host_path) fields. The [`plugin_id`](#plugin_id) and
[`constraint`](#constraint) fields are ignored. The [`node_pool`](#node_pool),
[`capacity_max`](#capacity_max), and [`capacity_min`](#capacity_min) fields are
ignored but must be consistent if set; the `node_pool` must match the node set
by `node_id`, and the minimum capacity must be less than the `capacity_max` and
`capacity`.

## Volume Placement

The `volume create` command creates the volume on a single node. If `node_id` is
set, the volume is created on that node. The `node_pool` must be unset or match
that node. Otherwise, the Nomad server iterates over the available nodes and
place the volume on the first available node based on the following factors:

* The node cannot already have a host volume with the same name.
* If `node_pool` is set, the selected node must be in that node pool.
* The node must meet any and all constraints defined by the `constraint` fields.

## Updating a Volume Definition

The `volume create` and `volume register` commands allow updating a volume
definition. However, after volume registration, you are only allowed to update
the following fields:

> * The `plugin_id` can be updated.
> * The `capacity_min` and `capacity_max` fields can be updated, and may increase
>   the volume size if the plugin supports it. Expansion may or may not be
>   possible while the volume is in use, again depending on the plugin.  Reducing
>   volume capacity is not allowed. Only available for volume creation.
> * The `capacity` field can be updated only for volume registration.
> * The `constraint` fields can be updated but are ignored after the volume is
>   created.
> * The `capability` blocks can be added or removed, but only if the capability is
>   not currently in use by a mounted volume.

Reviewer 2: ```suggestion
- `plugin_id`
- `capacity_min` and `capacity_max`. You may increase
  the volume size if the plugin supports it. Expansion may or may not be
  possible while the volume is in use, again depending on the plugin.  Reducing
  volume capacity is not allowed. Only available for volume creation.
- `capacity`, but only updated for volume registration.
- `constraint` fields. Ignored after the volume is created.

Additionally, you may add or remove `capability` blocks, but only if the
capability is not currently in use by a mounted volume.
```

Update to match the format in the CSI page. Unfortunately, I cannot figure out how to **not** render the field as a link. I don't know why/where DevDot is doing that. Doesn't matter if the list uses "-" or "*". I think it's weird that a list item that starts with an inline code block is rendered as a link to nowhere, as in the parameter lists. shrug.

> The `name`, `type`, and `node_id` fields cannot be updated. The `node_pool`
> field can only be changed from empty to the node pool that matches the `node_id`
> field.

Reviewer 2: ```suggestion
You cannot update the `name`, `type`, and `node_id` fields. You may only change the  `node_pool`
field from empty to the node pool that matches the `node_id`
field.
```

This is an example file used for the [`volume create`][] command.

```hcl
name            = "database"
type            = "host"
plugin_id       = "lvm-thin-provisioner"
node_pool       = "prod"

capacity_min = "80G"
capacity_max = "100G"

capability {
  access_mode     = "single-node-reader-only"
  attachment_mode = "file-system"
}

capability {
  access_mode     = "single-node-writer"
  attachment_mode = "file-system"
}

parameters {
  skuname = "Premium_LRS"
}
```

### Volume registration

This is an example file used for the [`volume register`][] command.

```hcl
name      = "database"
type      = "host"
node_id   = "a7b4c0ca-cc78-11ef-8b5a-cb6ea67b844c"
host_path = "/var/srv/example"
capacity  = "80G"

capability {
  access_mode     = "single-node-reader-only"
  attachment_mode = "file-system"
}

capability {
  access_mode     = "single-node-writer"
  attachment_mode = "file-system"
}

parameters {
  skuname = "Premium_LRS"
}
```

### Example Volume Expansion

The [`volume create`][] command can trigger an expansion to occur, after the
volume has already been created or registered.

Example using `volume create`, with a `volume.hcl` file:

```hcl
name      = "database"

> type      = "host"
> plugin_id = "external-plugin"
> 
> capacity_min = "30GiB"
> capacity_max = "50GiB"
> ```
> 
> Next, use the [`volume create`][] command and the `volume.hcl` file to create
> the volume:
> 
> ```shell-session
> $ nomad volume create ./volume.hcl
> ==> Created host volume database with ID 0c903229-311d-ba8a-f77e-45c31b83fab3
>   ✓ Host volume "0c903229" ready
> 
>     2025-01-06T16:56:09-05:00
>     ID        = 0c903229-311d-ba8a-f77e-45c31b83fab3
>     Name      = database
>     Namespace = default
>     Plugin ID = external-plugin
>     Node ID   = cfe033a7-50de-2c46-cd18-12be7429eeb3
>     Node Pool = default
>     Capacity  = 50 GiB
>     State     = ready
>     Host Path = /run/nomad/dev/alloc_mounts/0c903229-311d-ba8a-f77e-45c31b83fab3
> ```
> 
> Review the volume's current capacity with the [`volume status`][] command.
> 
> ```shell-session
> $ nomad volume status -type=host 0c903229 | grep Capacity
> Capacity             = 50 GiB
> ```
> 
> Next, to increase volume capacity, update the `capacity_min` and `capcity_max`
> fields in the `volume.hcl` file.
> 
> ```hcl
> id        = "0c903229-311d-ba8a-f77e-45c31b83fab3"
> name      = "database"
> type      = "host"
> plugin_id = "external-plugin"
> 
> capacity_min = "100GiB" # double
> capacity_max = "100GiB" # increased to match
> ```
> 
> Run the `volume create` command to expand the volume.
> 
> ```shell-session
> $ nomad volume create volume.hcl
> Created host volume database with ID 0c903229-311d-ba8a-f77e-45c31b83fab3
> ```
> 
> Use the `volume status` command to make sure the volume has the new capacity.
> 
> ```shell-session
> $ nomad volume status -type=host 0c903229 | grep Capacity
> Capacity             = 100 GiB
> ```

Reviewer 2: ````suggestion
This example shows how to expand a volume after it has been created or
registered.

1. Create a volume configuration file called `volume.hcl` that defines name,
   type, plugin ID, and capacity.

  ```hcl
   name      = "database"
   type      = "host"
   plugin_id = "external-plugin"

   capacity_min = "30GiB"
   capacity_max = "50GiB"
   ```

1. Create the volume using the [`volume create`] command.

   ```shell-session
   $ nomad volume create ./volume.hcl
   ==> Created host volume database with ID 0c903229-311d-ba8a-f77e-45c31b83fab3
   ✓ Host volume "0c903229" ready

    2025-01-06T16:56:09-05:00
    ID        = 0c903229-311d-ba8a-f77e-45c31b83fab3
    Name      = database
    Namespace = default
    Plugin ID = external-plugin
    Node ID   = cfe033a7-50de-2c46-cd18-12be7429eeb3
    Node Pool = default
    Capacity  = 50 GiB
    State     = ready
    Host Path = /run/nomad/dev/alloc_mounts/0c903229-311d-ba8a-f77e-45c31b83fab3
   ```

1. Review the volume's current capacity using the [`volume status`] command.

   ```shell-session
   $ nomad volume status -type=host 0c903229 | grep Capacity
   Capacity             = 50 GiB
   ```

1. Increase volume capacity in the `volume.hcl` file.

   Update the `capacity_min` and `capacity_max` fields.

   ```hcl
   id        = "0c903229-311d-ba8a-f77e-45c31b83fab3"
   name      = "database"
   type      = "host"
   plugin_id = "external-plugin"

   capacity_min = "100GiB" # double
   capacity_max = "100GiB" # increased to match
   ```

1. Expand the volume using the [`volume create`] command.

   The `volume create` command can trigger an expansion to occur, after the volume has already been created or registered.

   ```shell-session
   $ nomad volume create volume.hcl
   Created host volume database with ID 0c903229-311d-ba8a-f77e-45c31b83fab3
   ```

1. Review the new capacity by running the [`volume status`] command.

   ```shell-session
   $ nomad volume status -type=host 0c903229 | grep Capacity
   Capacity             = 100 GiB
   ```
````

Reviewer 2: Example of volume expansion steps in a list.

![image](https://github.com/user-attachments/assets/373c814e-93e5-4452-8e0b-23891c83b652)
